enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+
+OLM will report the following upgradeable conditions based on the collection of CSV present on cluster:
+
+| Key                  | Meaning                                                                                            |
+| ---------------------| -------------------------------------------------------------------------------------------------- |
+| Upgradeable    CSV   | A CSV with a valid upgradeable annotation that will run on the next Minor version of OpenShift     |
+| Non-Upgradeable  CSV | A CSV with a valid upgradeable annotation that will not run on the next Minor version of OpenShift |
+| Undeterminable CSV   | A CSV that does not include a valid upgradeable annotation                                         |
+
+| Upgradeable CSV Present | Indeterminate CSV Present | Non-Upgradeable CSV Present | Status          | Message                                                                                            |
+| ----------------------- | ------------------------- | --------------------------- |---------------- | -------------------------------------------------------------------------------------------------- |
+| Yes                     | No                        | No                          | Upgradeable     | Ready for upgrade                                                                                  |
+| Don't Care              | Yes                       | No                          | Upgradeable     | The following operators may not run on the next OpenShift Version: namespace/foo, namespace/bar"   |
+| Don't Care              | Don't Care                | Yes                         | Not Upgradeable | The following operators will not run on the next OpenShift Version: namespace/foo, namespace/bar"  |
+
+#### Pros
+
+- This approach provides allows Operator Authors to specify the exact versions of OpenShift that their operator supports.
+
+#### Cons
+
+- This makes upgrade safety opt-in for each Operator, meaning that OLM cannot guarantee upgrade safety unless each installed operator specifies a maxOpenShiftVersion.
+- As new versions of OpenShift become available, Operator Authors must update their operator's maxOpenShiftVersion field.
+
+### User Stories
+
+#### Story 1
+
+As an OpenShift Cluster Admin, I want OLM to block cluster upgrades if it can determine that at least one operator will not run on the next Minor OpenShift version.
+
+### Risks and Mitigations
+
+- The most immediate risk with this solution is the fact that OLM will prevent the Cluster from upgrading if an OLM Managed Operator will not run on the next Minor OpenShift Version. This is easily circumvented if the Cluster Admin initiates the upgrade with the `force` command or by removing the offending Operator.
+
+## Design Details
+
+### Open Questions
+
+- How will OLM rejecting an upgrade be surfaced in both the UI and the CLI?
```

Comment:
Should be no need for anything special.  The message OLM sets in its ClusterOperator will be picked up by the CVO and stored in ClusterVersion.  And both `oc adm upgrade ...` and the web console expose ClusterVersion's `Upgradeable` condition today, and there's an alert for `Upgradeable=False`.

Reply:
Not sure which scenario this bullet was aimed at, but there's not just "rejecting" updates, there's also the case where we are warning the admin that we can't be sure (because one or more operators doesn't make any declaration at all).

Reply:
Rewording for clarity:

> - When a CSV is present that does not include a maxOpenShiftVersion annotation, how will this information be portrayed via the Console UI and CLI?